File: LeetCode/Note_0094_Binary_Tree_Inorder_Traversal/Note_0094_Binary_Tree_Inorder_Traversal.java
import java.util.*;


/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {

    List<Integer> output = new ArrayList<>();

    public static class TreeNode {
        int val;
        TreeNode left;
        TreeNode right;
        TreeNode() {}
        TreeNode(int val) { this.val = val; }
        TreeNode(int val, TreeNode left, TreeNode right) {
            this.val = val;
            this.left = left;
            this.right = right;
        }
    }

    public List<Integer> inorderTraversal(TreeNode root) {
        this.helper(root);
        return this.output;
    }

    public void helper(TreeNode root) {
        if (root == null) {
            return;
        }

        this.helper(root.left);
        this.output.add(root.val);
        this.helper(root.right);
    }

    public static void main(String[] args) {
        System.out.println("Note_0094_Binary_Tree_Inorder_Traversal!"); 
        TreeNode node2 = new TreeNode(3, null, null);
        TreeNode node1 = new TreeNode(2, node2, null);
        TreeNode root = new TreeNode(1, null, node1);
        
        Solution solution = new Solution();
        solution.inorderTraversal(root);
        System.out.println(solution.output);
    }
}

File: LeetCode/Note_0094_Binary_Tree_Inorder_Traversal/Note_0094_Binary_Tree_Inorder_Traversal.py
from typing import List
from typing import Optional

# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right
class Solution:
    def inorderTraversal(self, root: Optional[TreeNode]) -> List[int]:
        output = []
        def recursive(node):
            if node is None:
                return

            recursive(node.left)
            output.append(node.val)
            recursive(node.right)
            
        recursive(root)
        return output
   

if __name__ == '__main__':
    print("kai:Note_0094_Binary_Tree_Inorder_Traversal")
    root = TreeNode()
    root.val = 1

    node1 = TreeNode()
    node1.val = 2

    node2 = TreeNode()
    node2.val = 3

    root.right = node1
    node1.left = node2

    solution = Solution()
    print(solution.inorderTraversal(root))
    
# T:O(n)
# S:O(n)